openupgrade_scripts/scripts/hr_recruitment/18.0.1.1/post-migration.py

```diff
@@ -0,0 +1,172 @@
+from openupgradelib import openupgrade, openupgrade_merge_records
+
+from odoo import Command
+
+
+def _fill_hr_candidate(env):
+    # create helper
+    column_adds = [
+        ("hr.candidate", "old_applicant_id", "many2one"),  # helper
+    ]
+    openupgrade.add_columns(env, column_adds)
+    # Create candidate records from applicant
+    openupgrade.logged_query(
+        env.cr,
+        """
+        INSERT INTO hr_candidate (
+            partner_id, partner_name, email_from, email_cc, email_normalized,
+            partner_phone, partner_phone_sanitized, phone_sanitized,
+            employee_id, linkedin_profile, type_id, priority,
+            availability, color, message_bounce, active,
+            user_id, company_id, old_applicant_id,
+            create_date, write_date, create_uid, write_uid
+        )
+        SELECT partner_id, partner_name, email_from, email_cc, email_normalized,
+               COALESCE(partner_mobile, partner_phone),
+               COALESCE(partner_mobile_sanitized, partner_phone_sanitized),
+               phone_sanitized, emp_id, linkedin_profile, type_id, priority,
+               availability, color, message_bounce, TRUE,
+               user_id, company_id, id,
+               NOW(), NOW(), create_uid, write_uid
+        FROM hr_applicant
+        WHERE candidate_id IS NULL;
+        """,
+    )
+    # Update hr_applicant.candidate_id using helper old_applicant_id
+    openupgrade.logged_query(
+        env.cr,
+        """
+        UPDATE hr_applicant ha
+        SET candidate_id = hc.id
+        FROM hr_candidate hc
+        WHERE ha.id = hc.old_applicant_id
+        """,
+    )
+    # remove helper
+    openupgrade.logged_query(
+        env.cr,
+        """
+        ALTER TABLE hr_candidate
+        DROP COLUMN old_applicant_id
+        """,
+    )
+    # fill many2many field categ_ids
+    openupgrade.logged_query(
+        env.cr,
+        """
+        INSERT INTO hr_applicant_category_hr_candidate_rel (
+            hr_candidate_id, hr_applicant_category_id
+        )
+        SELECT ha.candidate_id, m2m.hr_applicant_category_id
+        FROM hr_applicant_hr_applicant_category_rel as m2m
+            JOIN hr_applicant AS ha ON ha.id = m2m.hr_applicant_id
+        ON CONFLICT DO NOTHING
+        """,
+    )
+    # Fill calendar_event.candidate_id
+    openupgrade.logged_query(
+        env.cr,
+        """
+        UPDATE calendar_event ce
+        SET candidate_id = ha.candidate_id
+        FROM hr_applicant ha
+        WHERE ce.applicant_id = ha.id
+        """,
+    )
+    # unify candidates
+    env.cr.execute(
+        """
+        SELECT array_agg(id order by create_date desc), partner_id
+        FROM hr_candidate
+        GROUP BY partner_id
+        HAVING COUNT(id) > 1
```

Comment:
I think you must filter only if the partner is set; otherwise, you will merge records with an empty partner, and this is incorrect.

```suggestion
        SELECT array_agg(id order by create_date desc), partner_id
        FROM hr_candidate
        WHERE partner_id IS NOT NULL
        GROUP BY partner_id
        HAVING COUNT(id) > 1
```